### PR TITLE
Make dataclasses dependency conditional at install time

### PR DESCRIPTION
**Pull request by tmct**
Ensures dataclasses is not treated as a dependency of the "any" wheel on Python 3.7+

**PR type:** bugfix

**Related issue(s)/PRs:** -

## Summary

**Proposed changes**
* Make dataclasses dependency conditional at install time rather than conditional at wheel build time

**What alternatives have you considered?**
* Asking you to deprecate Python 3.6
* Asking you to create separate wheels for py36, py38 etc.

### Minimal working example

```bash
python --version  # 3.8
pip install gpflow
# dataclasses has been installed into site-packages! :(
```

### Release notes

**Fully backwards compatible:** yes

**If not, why is it worth breaking backwards compatibility:**

## PR checklist

- [ ] New features: code is well-documented
  - [ ] detailed docstrings (API documentation)
  - [ ] notebook examples (usage demonstration)
- [ ] The bug case / new feature is covered by unit tests
- [ ] Code has type annotations
- [ ] Build checks
  - [ ] I ran the black+isort formatter (`make format`)
  - [ ] I locally tested that the tests pass (`make check-all`)
- [ ] Release management
  - [ ] RELEASE.md updated with entry for this change
  - [ ] New contributors: I've added myself to CONTRIBUTORS.md

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -33,12 +33,10 @@
         "typing_extensions",
         "packaging",
         "deprecated",
+        "dataclasses; python_version < "3.7",
```
Comment (jesnie) [inline setup.py:36]: This doesn't look right. Maybe something like this:
```suggestion
        "dataclasses; python_version < '3.7'",
```
Our continuous integration tools are also complaining.
Maybe you can test this on your local machine?

**Review by jesnie — COMMENTED** [review]
Thanks for contributing!

Feel free to add yourself to the contributors list, and update the release notes while you're at it, so everybody can know about the hard work you are doing! :-)

File: setup.py (hunk 1 above)
Comment (tmct) [reply setup.py:36]: Apologies! I will correct and run the test locally

Comment (jesnie) [reply setup.py]: This looks much better. Thanks!

**Review by jesnie — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**